### PR TITLE
Recipe Generation V1, Even More QOL and Bug Fixes

### PR DESCRIPTION
**Pull request by RapidThrower264**
**Recipe Generation**
- Created a new command that allows the ability to create a 3x3 Minecraft recipe which can have various items put on them.
- Can be used through the `/gen recipe` command and putting in a string of the recipe in the following format
`<slot number>,<amount>,<item name (or skull)>,<extra details>`
Some extra details include changing leather armor color (Hex format), enchantment glint (`enchant`).
image

(This is limited to only items and skulls for the moment because I don't know of a good sprite sheet).

**QOL**
- Allow the ability to directly copy NBT from in game into the command.
- Added the ability to display Minecraft items (including skulls) along side the parsed item using the command options.
- Added the ability to display Minecraft items (including skulls) along side full generations using the command options.
- Added the ability to display both items and recipes along side full generations using the command options.
- Added the usage of `%%<color/style>:<text>%%` to allow people to set a small portion of text to another color.
   - `Reduce damage by %%RED:10% %%for %%GREEN:10s%% after casting the ability.` gets converted into 
   `Reduce damage by %%RED%%10% %%GRAY%%for %%GREEN%%10s%%GRAY%% after casting the ability`
- Using text generation will no longer put the extra space between the first two lines.
- Renamed All `itemgen` commands to `gen`.

**Font File Changes**
- Added Ⓑ.

**Bug Fixes**
- Properly fix the bug that causes errors on Line Lengths set between 1-4.
- Fix italics, bold, strikethrough and underlines not carrying over styles.
- Fix line length being one less than it actually is.

### CODE REVIEW

**Review by Aerhhh — CHANGES_REQUESTED** [review]

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    public static final String DESC_MAX_LINE_LENGTH = "Sets the maximum length for a line (1 - " + StringColorParser.MAX_FINAL_LINE_LENGTH + ") default " + StringColorParser.MAX_STANDARD_LINE_LENGTH;
+    public static final String DESC_ITEM_NAME = "The name of the Item stack you want to display";
+    public static final String DESC_HEAD_ID = "The ID of the skin or the Player Name (set is_player_name to True if it is a player name)";
+    public static final String DESC_IS_PLAYER_NAME = "If the skin ID given describes the player's name";
+    public static final String DESC_HIDDEN = "If you only want the generated image visible to yourself";
+    public static final String DESC_PARSE_ITEM = "Item JSON Display Data (in the form {\"Lore\": [...], \"Name\": \"\"}";
+    public static final String DESC_INCLUDE_ITEM = "Include the item along with the parsed description";
+    public static final String DESC_RENDER_INVENTORY = "If the Minecraft Inventory background should be drawn";
+    public static final String DESC_RECIPE = "The recipe for crafting the item (/gen recipe for more info)";
```
[the remark this answers is not in the document]
Comment (RapidThrower264) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java:35]: The description for commands must be constant expressions so I can't use `.formatted()`.

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    public static final String ITEM_INFO_ARGUMENTS = """
+                        `name`: The name of the item. Defaults to the rarity color, unless the rarity is none.
+                        `rarity`: Takes any SkyBlock rarity. Can be left as NONE.
+                        `item_lore`: Parses a description, including color codes, bold, italics, and newlines.
+                        `type`: The type of the item, such as a Sword or Wand. Can be left blank.
+                        `disable_rarity_linebreak (true/false)`: To be used if you want to disable automatically adding the empty line between the item lore and rarity.
+                        `alpha`: Sets the transparency of the background layer. 0 for transparent, 255 for opaque (default). 245 for overlay.
+                        `padding`: Adds transparency around the entire image. Must be 0 (default) or higher.
+                        `max_line_length`: Defines the maximum length that the line can be. Can be between 1 and 54.
+                        """;
```
[the remark this answers is not in the document]
Comment (RapidThrower264) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java:56]: Could do, but some of them are larger than 100 characters and provide a bit more information than what is used for the command option.

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorBuilder.java [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+
+    /***
+     * Creates a rendered Minecraft Inventory with up to 9 different slots in a 3x3 grid
+     * @param event the event associated to the command
+     * @param recipeString the string which contains the recipe items
+     * @return a rendered minecraft crafting recipe
+     */
+    @Nullable
+    public BufferedImage buildRecipe(GuildSlashEvent event, String recipeString, boolean renderBackground) {
+        // checking that the resources were correctly
```
[the remark this answers is not in the document]
Comment (RapidThrower264) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorBuilder.java:358]: Correctly leaving in a spaceship to Mars.

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aerhhh) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java]: Don't you do it on line 46 though?

[the remark this answers is not in the document]
Comment (RapidThrower264) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java]: Line 46 is a string that is sent to the user after the bot has been initialised, whereas DESC_RECIPE has to be ready in compilation, in which `.formatted()` isn't run.

(changed the string to read `The recipe for crafting the item (see recipe_help for more info)`)

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorBuilder.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+                items.put(item.getName(), item);
+                if (item.getOverlay() != null) {
+                    (item.getOverlay().acceptedParameters() == 1 ? oneParameterItems : twoParameterItems).append(item.getName()).append(", ");
+                }
+            }
+
+            oneParameterItems.delete(oneParameterItems.length() - 2, oneParameterItems.length());
+            twoParameterItems.delete(twoParameterItems.length() - 2, twoParameterItems.length());
+            GeneratorStrings.RECIPE_INFO_OTHER_INFORMATION = "Ability to Change One Layer (One Hex Color Parameter)\n" +
+                    oneParameterItems + "\n\nAbility to Change Both Layers (Two Hex Color Parameters)\n" + twoParameterItems;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/GeneratorBuilder.java:97]: Make these bold and don't capitalise the Change X Layers part

**Review by Aerhhh — CHANGES_REQUESTED** [review]

File: src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aerhhh) [reply src/main/java/net/hypixel/nerdbot/generator/GeneratorStrings.java]: It seems to work fine for me so it can probably be changed

**Pull request merged**